normandy/studies/models.py

```diff
@@ -131,4 +135,11 @@ def save(self, *args, **kwargs):
             if "xpi" in dirty_field_names:
                 self.populate_metadata()
 
+                filename = get_extension_filename(self)
+
+                if Extension.objects.filter(xpi=filename).exclude(pk=self.pk).exists():
+                    # Django just renames the file with a suffix if there is collision and the
+                    # FileField is set to `unique=True`. This forces an integrity error.
```

Comment:
That's horrible that Django does that! Good catch!

It always makes me nervous when you take an exception class from somewhere else and pretend it. I think it exists in `django.db` so you can use it for catching. Have you considered making your own exception class? A custom one?

Reply:
Hey, drive by comment because I saw this in my email. Is this a TOCTOU bug waiting to happen? (It may not be a problem even if it is...)

Reply:
> Have you considered making your own exception class? A custom one?

I had not considered that. This is the exception I expect to be thrown when there is a constraint violation so it seemed appropriate to raise it here.

Reply:
> Is this a TOCTOU bug waiting to happen?

That's a good point, it may be. I'm not sure if that really matters but I'll do some digging into a way to avoid that. I've been toying with the idea of changing the storage backend to error on overwrite.

Reply:
> I've been toying with the idea of changing the storage backend to error on overwrite.

I like this option a lot.